### PR TITLE
[RORDEV-352] remote reindex issue fix

### PR DESCRIPTION
**Pull request by pielas**
🐞Fix (ES) Fixed remote reindex operation with "type" parameter

### CODE REVIEW

File: es55x/src/main/scala/tech/beshu/ror/es/request/context/types/ReindexEsRequestContext.scala [hunk 1]
```diff
@@ -42,46 +42,25 @@ class ReindexEsRequestContext(actionRequest: ReindexRequest,
...
-        case Some(destinationIndex) if indices.toList.contains(destinationIndex) =>
-          Modified
-        case Some(_) | None =>
-          logger.info(s"[${id.show}] Destination index of _reindex request is forbidden")
-          ShouldBeInterrupted
-      }
+    if (isDestinationIndexOnFilteredIndicesList && isSearchRequestComposedOnlyOfAllowedIndices) {
+      Modified
+    } else {
+      logger.info(s"[${id.show}] Destination index of _reindex request is forbidden")
```
Comment (coutoPL) [inline es55x/src/main/scala/tech/beshu/ror/es/request/context/types/ReindexEsRequestContext.scala:62]: looks like we should recognise two cases:
1. destination index is forbidden
2. at least one index from sources indices list is forbidden

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala [hunk 2]
```diff
@@ -40,13 +40,25 @@ trait ReindexSuite
   "A reindex request" should {
     "be able to proceed" when {
       "user has permission to source index and dest index"  in {
-        val result = user1ActionManager.actionPost("_reindex", ReindexSuite.reindexPayload("test1_index"))
+        val result = user1ActionManager.actionPost("_reindex", ReindexSuite.reindexPayload("test1_index", "test1_index_reindexed"))
         assertEquals(200, result.getResponseCode)
```
Comment (coutoPL) [inline integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala:44]: btw. let's refactor it and use scalatest matchers instead

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala [hunk 3]
```diff
@@ -40,13 +40,25 @@ trait ReindexSuite
...
       }
     }
     "not be able to proceed" when {
-      "user has no permission to source index and dest index"  in {
-        val result = user1ActionManager.actionPost("_reindex", ReindexSuite.reindexPayload("test2_index"))
+      "user has no permission to source index and dest index which are present on ES"  in {
+        val result = user1ActionManager.actionPost("_reindex", ReindexSuite.reindexPayload("test2_index", "test2_index_reindexed"))
```
Comment (coutoPL) [inline integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala:49]: let's refactor also this one:
* don't use actionPost (we'll remove ActionManagerJ in the future)
* add proper method (more specialised one) to IndexManager

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala [hunk 4]
```diff
@@ -40,13 +40,25 @@ trait ReindexSuite
...
+      }
+      "user has no permission to source index and dest index which are absent on ES"  in {
+        val result = user1ActionManager.actionPost("_reindex", ReindexSuite.reindexPayload("not_allowed_index", "not_allowed_index_reindexed"))
+        assertEquals(401, result.getResponseCode)
+      }
+      "user has permission to source index and but no permission to dest index"  in {
+        val result = user1ActionManager.actionPost("_reindex", ReindexSuite.reindexPayload("test1_index", "not_allowed_index_reindexed"))
+        assertEquals(401, result.getResponseCode)
+      }
+      "user has permission to dest index and but no permission to source index"  in {
```
Comment (coutoPL) [inline integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala:60]: and one more: user has no access to one of requested source indices

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteReindexSuite.scala [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+      rorContainerSpecification = ContainerSpecification(
+        Map(
+          "reindex.remote.whitelist" -> "*:9200",
+          "reindex.ssl.verification_mode" -> "none"
+        )),
+      xPackSupport = false
+    )(rorConfigFileName)
+  )
+
+  private lazy val sourceEsActionManager = new ActionManagerJ(clients.head.basicAuthClient("dev1", "test"))
```
Comment (coutoPL) [inline integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteReindexSuite.scala:58]: same as in previous suite

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteReindexSuite.scala [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  private lazy val destEsActionManager = new ActionManagerJ(clients.last.basicAuthClient("dev1", "test"))
+
+  private lazy val sourceEsContainer = sourceEsCluster.nodes.head
+  private lazy val destEsContainer = destEsCluster.nodes.head
+
+  private lazy val oldEsContainerAddress = s"https://${sourceEsContainer.containerInfo.getConfig.getHostName}:9200"
```
Comment (coutoPL) [inline integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteReindexSuite.scala:64]: shouldn't we add this method to EsContainer?

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteReindexSuite.scala [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+  lazy val clusterContainers: NonEmptyList[EsClusterContainer] = NonEmptyList.of(sourceEsCluster, destEsCluster)
+  lazy val esTargets: NonEmptyList[EsContainer] = NonEmptyList.of(sourceEsCluster.nodes.head, destEsCluster.nodes.head)
+
+  "A remote reindex request" should {
+    "be able to proceed" when {
+      "request specifies source and dest index that are allowed on both source and dest ES" in {
+        val result = destEsActionManager.actionPost("_reindex", RemoteReindexSuite.remoteReindexPayload("test1_index", "test1_index_reindexed", "dev1", oldEsContainerAddress))
+        assertEquals(200, result.getResponseCode)
```
Comment (coutoPL) [inline integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteReindexSuite.scala:73]: same here

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteReindexSuite.scala [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      }
+    }
+    "be blocked by dest ES" when {
+      "request specifies source index that is allowed, but dest that isn't allowed" in {
+        val result = destEsActionManager.actionPost("_reindex", RemoteReindexSuite.remoteReindexPayload("test1_index", "not_allowed_index","dev1", oldEsContainerAddress))
```
Comment (coutoPL) [inline integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteReindexSuite.scala:78]: also add specialised method to IndexManager

**Review by coutoPL — COMMENTED** [review]
+ it seems that we should describe the _reindex case in docs right?

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala (hunk 4 above)
Comment (pielas) [reply integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala:60]: Do you mean request with multiple source indices? Is it even possible?

**Comment by pielas** [comment]
> * it seems that we should describe the _reindex case in docs right?

Functionality hasn't been changed much, and it's pretty straightforward, but I can describe it if you think it's needed

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala (hunk 4 above)
Comment (coutoPL) [reply integration-tests/src/test/scala/tech/beshu/ror/integration/suites/ReindexSuite.scala:64]: no, it isn't. You are right.

**Review by coutoPL — APPROVED** [review]
LGTM! Thanks

**Pull request merged**